pyprocessing/math.py

```diff
@@ -162,191 +179,214 @@ def from_angle(angle, target=None):
-        self._z = value
-        if value is not None:
-            self.is_2d = False
+    @property
+    def is_3d(self):
+        return len(self) == 3
 
     @property
-    def position(self):
-        return self.x, self.y, self.z
+    def is_4d(self):
+        return len(self) == 4
 
     def set(self, *vector, x=None, y=None, z=None):
-        if vector:
-            if isinstance(vector[0], PVector):
-                self.x, self.y, self.z = vector[0].position
-            elif isinstance(vector[0], (list, tuple)):
-                self.x, self.y, *_ = vector[0]
-                if len(vector[0]) == 3:
-                    self.z = vector[0][2]
-                else:
-                    self.z = None
-            elif all(isinstance(p, (int, float)) for p in vector):
-                self.x, self.y, self.z = vector
-            elif isinstance(vector[0], complex):
-                self.x, self.y, self.z = vector[0].real, vector[0].imag, None
-        if x is not None:
-            self.x = x
-        if y is not None:
-            self.y = y
-        if z is not None:
-            self.z = z
+        self.seq = PVector(*vector, x, y, z).seq
 
     def copy(self):
-        return PVector(self.x, self.y, self._z)
+        return PVector(*self)
 
     def mag(self):
         return math.sqrt(self.mag_sq())
 
     def mag_sq(self):
-        return self.x * self.x + self.y * self.y + self.z * self.z
+        return sum(component_a * component_a for component_a in self)
 
     def dot(self, other):
-        return self.x * other.x + self.y * other.y + self.z * other.z
+        return sum(component_a * component_b for component_a, component_b in zip(self, other))
 
     def cross(self, other):
-        x = self.y * other.z - self.z * other.y
-        y = self.z * other.x - self.x * other.z
-        z = self.x * other.y - self.y * other.x
+        if not len(self) == 3:
+            raise ValueError(
+                'Cross product is only defined for 3d vectors'
+            )
+        x = self[1] * other[2] - self[2] * other[1]
+        y = self[2] * other[0] - self[0] * other[2]
+        z = self[0] * other[1] - self[1] * other[0]
         return PVector(x, y, z)
 
     def normalize(self):
-        mag = self.mag()
-        self.x /= mag
-        self.y /= mag
-        self.z /= mag
+        s = self.mag()
+        if s == 0:
+            return
+        for i in range(len(self)):
+            self[i] /= s
+
+    def normalized(self):
+        v = self.copy()
+        v.normalize()
+        return v
 
     def limit(self, limit):
-        if self.mag_sq() <= limit ** 2:
+        s = self.mag_sq()
+        if s <= limit * limit:
             return
 
-        self.normalize()
-        self *= limit
+        s = math.sqrt(s) / limit
+
+        for i in range(len(self)):
+            self[i] /= s
 
     def heading(self):
-        if not self.is_2d:
-            raise ValueError('Cannot calculate heading on a 3D Vector.')
```

Comment:
don't remove that. Technically, you can calculate the heading by projecting onto a plane, but since no plane is defined, we assume it's just for the XY plane, thus, only for 2D vectors. The message can be changed to reflect that

Reply:
Sure, I forget we can do `vector.xy.heading()`